spec/compiler/apply_types_spec.cr

```diff
@@ -0,0 +1,411 @@
+    run_source_typer_spec(<<-INPUT, <<-OUTPUT, named_splats: false)
+    def hello(*arg, **args)
+      nil
+    end
+    hello(2, hello: "world", world: 3)
+    INPUT
+    def hello(*arg : Int32, **args) : Nil
+      nil
+    end
+
+    hello(2, hello: "world", world: 3)
+    OUTPUT
+  end
+
+  it "types double plats but not splats" do
+    run_source_typer_spec(<<-INPUT, <<-OUTPUT, splats: false)
+    def hello(*arg, **args)
+      nil
+    end
+    hello(2, hello: "world", world: 3)
+    INPUT
+    def hello(*arg, **args : String | Int32) : Nil
+      nil
+    end
+
+    hello(2, hello: "world", world: 3)
+    OUTPUT
+  end
+
+  it "doesn't type double splat with empty call" do
+    run_source_typer_spec(<<-INPUT, <<-OUTPUT)
+    def hello(**args)
+      nil
+    end
+    hello(hello: "world", world: 3)
+    hello
+    INPUT
+    def hello(**args) : Nil
+      nil
+    end
+
+    hello(hello: "world", world: 3)
+    hello
+    OUTPUT
+  end
+
+  it "types blocks" do
+    run_source_typer_spec(<<-INPUT, <<-OUTPUT)
+    def hello(&block)
+      block
+    end
+    hello {}
+    INPUT
+    def hello(&block : Proc(Nil)) : Proc(Nil)
+      block
+    end
+
+    hello { }
+    OUTPUT
+  end
+
+  it "types class instance methods" do
+    run_source_typer_spec(<<-INPUT, <<-OUTPUT, line_number: 2)
+    class Test
+      def hello(arg)
+        arg
+      end
+    end
+    Test.new.hello(3)
+    INPUT
+    class Test
+      def hello(arg : Int32) : Int32
+        arg
+      end
+    end
+
+    Test.new.hello(3)
+    OUTPUT
+  end
+
+  it "types class methods" do
+    run_source_typer_spec(<<-INPUT, <<-OUTPUT, line_number: 2)
+    class Test
+      def self.hello(arg)
+        arg
+      end
+    end
+    Test.hello(3)
+    INPUT
+    class Test
+      def self.hello(arg : Int32) : Int32
+        arg
+      end
+    end
+
+    Test.hello(3)
+    OUTPUT
+  end
+
+  it "types method included from module" do
+    run_source_typer_spec(<<-INPUT, <<-OUTPUT, line_number: 2)
+    module IncludeMe
+      def hello(arg)
+        arg
+      end
+    end
+
+    class Test
+      include IncludeMe
+    end
+    Test.new.hello(3)
+    INPUT
+    module IncludeMe
+      def hello(arg : Int32) : Int32
+        arg
+      end
+    end
+
+    class Test
+      include IncludeMe
+    end
+
+    Test.new.hello(3)
+    OUTPUT
+  end
+
+  it "doesn't remove newline when inserting return types" do
+    run_source_typer_spec(<<-INPUT, <<-OUTPUT)
+    def hello
+      # world
+      "world"
+    end
+    hello
+    INPUT
+    def hello : String
+      # world
+      "world"
+    end
+
+    hello
+    OUTPUT
+  end
+
+  it "turns unions with nil to have a '?' suffix" do
+    run_source_typer_spec(<<-INPUT, <<-OUTPUT)
+    def hello(arg)
+      nil
+    end
+    hello(nil)
+    hello("world")
+    INPUT
+    def hello(arg : String?) : Nil
+      nil
+    end
+
+    hello(nil)
+    hello("world")
+    OUTPUT
+  end
+
+  # it "parses, runs semantic, and types everything" do
```

Comment:
Commenting out this spec for now - it runs prelude as well as try to type everything the test specific file (but that test file isn't part of this PR yet).